### PR TITLE
perilog: support direct execution of prolog/epilog on job ranks for better performance

### LINKED ISSUE #6216: `flux perilog-run` is not scalable
Problem: For large jobs, the `flux perilog-run --exec-per-rank` solution for executing `prolog` and `epilog` actions on each node of spends most of its time executing processes. It is estimated to take several minutes to run even around 5000 `flux exec` processes.

In contrast, flux-exec across 5000 ranks takes less than 1 second.

### PR DESCRIPTION
**Pull request by grondo**
Posting this PR as a WIP to get early feedback and to give reviewers a head start.

This PR reimplements the perilog jobtap plugin using bulk-exec in order to support direct execution of prolog/epilog on the ranks of a job instead of using `perilog-run`, which has been shown to take several minutes to start `flux exec` for a job comprising thousands of nodes.

The plugin should continue to work the same unless a new `per-rank = true` option is set, in which case the plugin uses bulk-exec to launch the defined `command` across all ranks that are part of the starting/ending job.

Support for other `flux perilog-run` features are also moved directly into the plugin, including draining failed ranks, a `timeout` parameter, and a `use-imp` boolean which replaces `command` with `["$imp", "run", "$name"]` if used.

The plugin also now supports a jobtap query callback which can be used to observe the current config, as well as active job prolog and epilogs, including which ranks are still actively running these proceses.

This will be a WIP because it still needs testing (results of which may change some of the code). 

Fixes #6216

### CODE REVIEW

**Comment by garlick** [comment]
I've built packages based on the PR pushed last night (2ff65286613f8afe96a267476a82c0f4bcf8b5ad) and have installed them on my test cluster.  I'll try to do some testing this morning.

**Comment by grondo** [comment]
Thanks! This morning I pushed an update to fix the error logging. I'll address those CodeQL alerts as well and then start in on writing some sharness tests.

**Comment by grondo** [comment]
FYI - just pushed a first cut at some testing.

File: src/modules/job-manager/Makefile.am
Comment (wihobbs) [inline src/modules/job-manager/Makefile.am:112]: Nit: commit message spelling "useful" "efficient"

File: src/modules/job-manager/plugins/perilog.c
Comment (wihobbs) [inline src/modules/job-manager/plugins/perilog.c]: Commit message "per-task" should be "per-rank"? Sorry if I'm misreading that, I just don't see any references to tasks in the code.

File: src/modules/job-manager/plugins/perilog.c
Comment (wihobbs) [inline src/modules/job-manager/plugins/perilog.c:106]: Commit message seems truncated?

**Review by wihobbs — COMMENTED** [review]
Just commit nits. :-)

File: src/modules/job-manager/plugins/perilog.c
Comment (grondo) [reply src/modules/job-manager/plugins/perilog.c:106]: Just an extra stray line I think.

**Comment by grondo** [comment]
Thanks @wihobbs! That's helpful. I've fixed the commit messages and will force-push.

File: src/modules/job-manager/plugins/perilog.c
Comment (garlick) [inline src/modules/job-manager/plugins/perilog.c:621]: Leaks `rl` on `rlist_ranks()` failure

File: src/modules/job-manager/plugins/perilog.c
Comment (garlick) [inline src/modules/job-manager/plugins/perilog.c:115]: commit message: s/doubel/double/

File: src/modules/job-manager/plugins/perilog.c
Comment (garlick) [inline src/modules/job-manager/plugins/perilog.c]: Minor: mention the switch to `perilog_proc_name()` in the commit description or perhaps split the commit

File: src/modules/job-manager/plugins/perilog.c
Comment (garlick) [inline src/modules/job-manager/plugins/perilog.c:101]: commit message: s/prolgo/prolog/

File: src/modules/job-manager/plugins/perilog.c
Comment (garlick) [inline src/modules/job-manager/plugins/perilog.c:515]: If perilog is unloaded with this future pending, seems like the job manager reactor could end up calling a bad address?

Should the future be part of the proc struct so it can be destroyed if it is in progress when perilog is unloaded?

File: src/modules/job-manager/plugins/perilog.c
Comment (garlick) [inline src/modules/job-manager/plugins/perilog.c]: Is there a way to disable the prolog timeout?  Seems like setting it to 0 would cause it to trigger immediately.  Maybe that should be special cased to "disabled"?

File: src/modules/job-manager/plugins/perilog.c
Comment (garlick) [inline src/modules/job-manager/plugins/perilog.c:357]: `hosts` could be null if both `flux_hostmap_lookup()` and `strdup()` fail

File: src/modules/job-manager/plugins/perilog.c
Comment (garlick) [inline src/modules/job-manager/plugins/perilog.c:156]: extra space bewteen `*` annd function name

**Review by garlick — COMMENTED** [review]
Nice!  There's some great changes here and this will probably have a big impact at scale.

I added some comments while going through the diffs sequentially but I think possibly some of my comments are off b/c this is still a WIP and includes some incremental development.  Ignore those please :-)

File: src/modules/job-manager/plugins/perilog.c
Comment (grondo) [reply src/modules/job-manager/plugins/perilog.c:585]: It is only passed to the log function after being checked if the line matches (or doesn't match) one of the log-ignore regexes. Since this function requires a NUL terminated buffer, it has to be copied.

**Comment by grondo** [comment]
The incremental develepment will probably stay and reflects how the features were developed. It would unfortunately be very time consuming to squash some of the development together, so it is either this or one big "rewrite perilog plugin" commit. I'm fine either way.

**Comment by garlick** [comment]
Ok by me if the incremental is intended.

**Comment by grondo** [comment]
I was waiting on removal of WIP once we get some feedback that the overall approach is acceptable. I was waiting before writing/updating any documentation to make sure everyone is fine with the configuration and other external design parameters. But as far as I'm concerned I could remove WIP now as I didn't plan on doing much more work besides testing and fixing any fallout from that.

**Comment by garlick** [comment]
It seems like a great improvement to me and shouldn't be controversial
since using the new functionality requires a config change to opt in.

File: src/modules/job-manager/plugins/perilog.c
Comment (grondo) [reply src/modules/job-manager/plugins/perilog.c:581]: will change to `if (len <= 0`...

File: src/modules/job-manager/plugins/perilog.c
Comment (grondo) [reply src/modules/job-manager/plugins/perilog.c:515]: Yeah, good catch! Note that unloading the perilog plugin with outstanding prolog/epilog will result in jobs with `{prolog,epilog}-start` events and no corresponding `{prolog,epilog}-finish` event though..

File: src/modules/job-manager/plugins/perilog.c
Comment (grondo) [reply src/modules/job-manager/plugins/perilog.c]: The timer watcher is not created unless `timeout > 0.0`:
```c
    if (timeout > 0.0) {
        flux_watcher_t *w;
        if (!(w = flux_timer_watcher_create (flux_get_reactor (h),
                                             timeout,
                                             0.,
                                             timeout_cb,
                                             proc))) {
```
So setting it to 0 disables the timer. You are suggesting that the literal string "disabled" should be a synonym for `0`?

Oh, I guess since `fsd_parse_duration()` is used, `inf`, `infinity` or `INF` would also work. However, this would have to be special cased in the stats, so maybe I'll just make Infinity and alias for 0, i.e. disabled (if that helps)

Comment (garlick) [reply src/modules/job-manager/plugins/perilog.c]: 0 is fine.  I must have misread the code.   It seemed like 0 went straight into the timer watcher `after` value which means time out immediately

File: src/modules/job-manager/plugins/perilog.c
Comment (grondo) [reply src/modules/job-manager/plugins/perilog.c:357]: and printf will report "(null)" -- would `"unknown"` be better.

Comment (garlick) [reply src/modules/job-manager/plugins/perilog.c:357]: sorry, I keep forgetting that we decided that is portably safe now.  FIne as is then.

File: src/modules/job-manager/plugins/perilog.c
Comment (garlick) [inline src/modules/job-manager/plugins/perilog.c:162]: Suggestion: squash these consecutive commits down to one or two:
- 6a61e3a6a8e44c190faa85ab9eee7103d777869b perilog: add use-imp config parameter
- ad9f0c1b657c222ed74993ea540172536e88dbfe perilog: improve error message on failure
- ff1d8a6a3821490532a9860022a021dc7c4252dc  perilog: remove use-imp config

File: doc/man5/flux-config-job-manager.rst
Comment (garlick) [inline doc/man5/flux-config-job-manager.rst:128]: Should `kill-timeout` be an FSD to match `timeout`?

File: doc/man5/flux-config-job-manager.rst
Comment (garlick) [inline doc/man5/flux-config-job-manager.rst:131]: Maybe "whether a fatal job exception raised while the prolog is active terminates the prolog" would be slightly clearer?
Meh, maybe not.  Up to you.

File: doc/guide/admin.rst
Comment (garlick) [inline doc/guide/admin.rst:678]: Might want to put the bit about the obsolete solution in a Note box.

**Review by garlick — APPROVED** [review]
This looks good, just a couple of minor comments.  I'll try to do a bit more testing this morning.

File: doc/man5/flux-config-job-manager.rst
Comment (grondo) [reply doc/man5/flux-config-job-manager.rst:128]: It could, but is mainly there for testing and it seemed like a bit of wasted effort because then it would require extra tests, etc. However, it is pretty simple if you feel strongly about it. Perhaps more important would be a note saying this parameter is mainly meant for testing?

File: doc/man5/flux-config-job-manager.rst
Comment (grondo) [reply doc/man5/flux-config-job-manager.rst:131]: Sure, that sounds better to me.

File: doc/guide/admin.rst
Comment (grondo) [reply doc/guide/admin.rst:678]: or should it just be removed?

File: src/modules/job-manager/plugins/perilog.c
Comment (grondo) [reply src/modules/job-manager/plugins/perilog.c:162]: Thanks for catching that, I had forgotten that incremental modification was in there.

**Comment by garlick** [comment]
Changes LGTM.

**Comment by garlick** [comment]
My admittedly light testing on my test cluster did not turn up any issues.  You have a green light from me :-)

**Comment by grondo** [comment]
Sheesh, I found a bunch of places where my fingers typed "per-task" when my brain said "per-rank". Fixed and forced a push.

**Comment by grondo** [comment]
Just as a data point, here's some timing differences between `per-rank` and `flux perilog-run` executing a prolog of `true` across an instance of size=720

`flux perilog-run`: 6.5s
```
0.055s: job.alloc {"annotations":{"sched":{"resource_summary":"rank[0-719]/core[0-47],gpu[0-7]"}}}
0.055s: job.prolog-start {"description":"job-manager.prolog"}
6.586s: job.prolog-finish {"description":"job-manager.prolog","status":0}
```

`per-rank = true`: 0.2 s
```
0.051s: job.alloc {"annotations":{"sched":{"resource_summary":"rank[0-719]/core[0-47],gpu[0-7]"}}}
0.057s: job.prolog-start {"description":"job-manager.prolog"}
0.276s: job.prolog-finish {"description":"job-manager.prolog","status":0}
0.485s: job.start
```

**Comment by grondo** [comment]
Ok, I haven't found any issues either, so I'll set MWP. Thanks!

**Pull request merged**
